src/python/pants/backend/python/rules/repl.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from pathlib import PurePath
+
+from pants.backend.python.rules.pex import Pex
+from pants.backend.python.rules.pex_from_target_closure import CreatePexFromTargetClosure
+from pants.base.build_root import BuildRoot
+from pants.engine.addressable import Addresses
+from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
+from pants.engine.legacy.graph import TransitiveHydratedTargets
+from pants.engine.legacy.structs import PythonTargetAdaptor
+from pants.engine.rules import goal_rule
+from pants.engine.selectors import Get
+from pants.option.global_options import GlobalOptions
+from pants.util.contextutil import temporary_dir
+
+
+logger = logging.getLogger(__name__)
+
+
+class PythonReplOptions(GoalSubsystem):
+  """Opens a REPL."""
+  name = 'repl2-python'
+
+
+class PythonRepl(Goal):
```

Comment:
I think that a reasonable amount of decoupling to begin heading in the direction of #9104 would be to install the `repl` goal in the core with a union, and to then fill the union in the python backend.